### PR TITLE
[1/5] Light client sync protocol

### PR DESCRIPTION
**Pull request by nicopernas**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

Implements the Altair spec logic defined in this version:

https://github.com/ethereum/consensus-specs/tree/208da34ac4e75337baf79adebf036ab595e39f15/specs/altair/light-client

It also implements the relevant endpoints specified at https://github.com/ethereum/beacon-APIs

https://github.com/prysmaticlabs/prysm/pull/11815

**Which issues(s) does this PR fix?**

Partially addresses https://github.com/prysmaticlabs/prysm/issues/11571

**Other notes for review**

This PR is a follow up to https://github.com/prysmaticlabs/prysm/pull/11815 which was too big to be reviewed. This is only the first PR in a series of 5 that I will be submitting sequentially. 

The heavy lifting here was done by [@qinlz2](https://github.com/qinlz2) and [@7AC](https://github.com/7AC), so kudos to them. 

-----
https://github.com/prysmaticlabs/prysm/pull/12854 is the last PR in the series that contains **all** changes so you can go back and see the full picture while reviewing this one.

### CODE REVIEW

**Review by potuz — DISMISSED** [review]
This looks good to me, requesting another approval cc @nisdas

File: proto/eth/v2/beacon_lightclient.proto [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+syntax = "proto3";
+
+package ethereum.eth.v2;
+
+import "proto/eth/ext/options.proto";
+import "proto/eth/v1/attestation.proto";
+import "proto/eth/v1/beacon_block.proto";
+import "proto/eth/v2/version.proto";
+import "proto/eth/v2/sync_committee.proto";
+//import "proto/engine/v1/execution_engine.proto";
```
Comment (terencechain) [inline proto/eth/v2/beacon_lightclient.proto:23]: Best to remove commented line if not used

```suggestion
```

File: proto/eth/v2/beacon_lightclient.proto [hunk 2]
```diff
@@ -0,0 +1,96 @@
+// Copyright 2021 Prysmatic Labs.
```
Comment (terencechain) [inline proto/eth/v2/beacon_lightclient.proto:1]:
```suggestion
// Copyright 2023 Prysmatic Labs.
```

**Comment by rkapka** [comment]
Hi @nicopernas . We are currently in the process of switching our APIs from gRPC to HTTP. The first step is this tracking issue: https://github.com/prysmaticlabs/prysm/issues/12635. We want to ultimately remove protobufs and use "pure" Go structs in our whole codebase, along with HTTP handlers serving API requests. That being said, we currently still use protobufs in our core code and for ssz generation.

I see that none of your types are added to the `ssz_gen_marshal` bazel rule, which makes me think that it is not necessary to define them as protos. On the other hand, all light client endpoints in the Beacon API specs allow an `Accept` header, meaning they can be returned in SSZ format. 

I think it would be most future-proof to split all types into 3 groups:
1. API-only structs defined in `/beacon-chain/rpc/beacon/light_client_structs.go`
2. Structs that will be used in core light client code and don't need SSZ serialization, defined under `/consensus-types/light-client` (there might not be any)
3. Structs that will be used in core light client code and need SSZ serialization, defined in `proto/eth/v2/beacon_lightclient.proto`

**Comment by nicopernas** [comment]
@rkapka thanks for your suggestion. 

Are you suggesting we remove the grpc endpoints and have the http layer serving requests directly? if so, which would be a good example to follow? Is this the right doc to follow https://hackmd.io/X9Epjx7ARMOXES0fye3ZRw#Reuse-code-from-beacon-chainprysmv1alpha1?

> API-only structs defined in /beacon-chain/rpc/beacon/light_client_structs.go

These are HTTP API structs I would imagine right? If so, we added them to`beacon-chain/rpc/apimiddleware/structs.go` in a later PR. See https://github.com/prysmaticlabs/prysm/pull/12854. I guess it's only a matter of moving them to where you suggested. These should not support SSZ marshalling right?

**Comment by rkapka** [comment]
> Are you suggesting we remove the grpc endpoints and have the http layer serving requests directly?

Yes

> if so, which would be a good example to follow?

https://github.com/prysmaticlabs/prysm/pull/12633/files is pretty simple. You basically have to implement the HTTP handler under `beacon-chain/rpc/eth/light-client` (I think that is a better separation then putting it under `/eth/beacon`) and register it in `beacon-chain/rpc/service.go`

> Is this the right doc to follow [hackmd.io/X9Epjx7ARMOXES0fye3ZRw#Reuse-code-from-beacon-chainprysmv1alpha1](https://hackmd.io/X9Epjx7ARMOXES0fye3ZRw#Reuse-code-from-beacon-chainprysmv1alpha1)?

This document is meant for a project where we rewrite **existing** grpc endpoints into HTTP, so it's not exactly what you want.

> These are HTTP API structs I would imagine right? If so, we added them tobeacon-chain/rpc/apimiddleware/structs.go in a later PR. See https://github.com/prysmaticlabs/prysm/pull/12854. I guess it's only a matter of moving them to where you suggested. These should not support SSZ marshalling right?

Right. `beacon-chain/rpc/eth/light-client/structs.go` will probably be the best place to put them. Using an HTTP handler means we can skip the middleware entirely, which makes things simpler and more performant. Also please note that requests/responses for HTTP handlers don't need the custom middleware tags, such as `hex="true"`.

**Comment by nicopernas** [comment]
@rkapka we had also added support for the http streaming API (see [this commit](https://github.com/prysmaticlabs/prysm/pull/12854/commits/8835bec722c186029d0fa703ec39e4d2bf2f8a61) for example) 
Is that something you are also planning to remove? That code path uses proto messages and grpc, so I'm thinking that you are either reworking it o removing it entirely?

**Comment by rkapka** [comment]
We will convert this to HTTP eventually, but extending it this way is fine for now.

**Comment by nicopernas** [comment]
I have updated this and the "last PR of the series" (https://github.com/prysmaticlabs/prysm/pull/12854) following your suggestions. I've removed the GRPC endpoints and replaced them with pure HTTP handlers. I've also got rid of most of the middleware changes and reduced the number of messages defined in the protobuf. 

Hopefully this aligns better with your plans and makes for an easier transition out of grpc :) 

If I missed some any of your guidance (which is very likely) please let me know and I will happily update the patch.

File: beacon-chain/rpc/eth/helpers/error_handling.go [hunk 3]
```diff
@@ -66,3 +68,38 @@ func PrepareStateFetchError(err error) error {
...
+	if err != nil {
+		return status.Errorf(codes.Internal, "Could not get block from block ID: %v", err)
+	}
+	if err := blocks.BeaconBlockIsNil(blk); err != nil {
+		return status.Errorf(codes.NotFound, "Could not find requested block: %v", err)
+	}
+	return nil
+}
+
+func HandleGetBlockErrorJson(blk interfaces.ReadOnlySignedBeaconBlock, err error) *http2.DefaultErrorJson {
```
Comment (rkapka) [inline beacon-chain/rpc/eth/helpers/error_handling.go:85]: Instead of creating this function, please use `WriteBlockFetchError` which is defined in `beacon-chain/rpc/eth/shared/errors.go`

File: beacon-chain/rpc/eth/helpers/lightclient/lightclient.go [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+func NewLightClientOptimisticUpdateFromBeaconState(
+	ctx context.Context,
+	config *params.BeaconChainConfig,
+	state state.BeaconState,
+	block interfaces.ReadOnlySignedBeaconBlock,
+	attestedState state.BeaconState) (*ethpbv2.LightClientUpdate, error) {
+	// assert compute_epoch_at_slot(attested_state.slot) >= ALTAIR_FORK_EPOCH
+	attestedEpoch := types.Epoch(uint64(attestedState.Slot()) / uint64(config.SlotsPerEpoch))
+	if attestedEpoch < types.Epoch(config.AltairForkEpoch) {
+		return nil, status.Errorf(codes.InvalidArgument, "Invalid attested epoch: %d", attestedEpoch)
```
Comment (rkapka) [inline beacon-chain/rpc/eth/helpers/lightclient/lightclient.go:29]: It would be better to return `RpcError` (defined in `beacon-chain/rpc/core/errors.go`) everywhere instead of `status` errors, which are a gRPC construct.

File: proto/eth/v2/custom.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
+package eth
+
+import (
+	"bytes"
+	"math/bits"
+)
+
+const (
+	NextSyncCommitteeIndex = uint64(55)
+	FinalizedRootIndex     = uint64(105)
```
Comment (rkapka) [inline proto/eth/v2/custom.go:10]: do these consts have to be exported?

File: proto/eth/v2/custom.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+	}
+	for i := range x.Pubkeys {
+		if !bytes.Equal(x.Pubkeys[i], other.Pubkeys[i]) {
+			return false
+		}
+	}
+	return bytes.Equal(x.AggregatePubkey, other.AggregatePubkey)
+}
+
+func FloorLog2(x uint64) int {
```
Comment (rkapka) [inline proto/eth/v2/custom.go:25]: does this function have to be exported?

File: proto/eth/v2/beacon_lightclient.proto [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    v1.BeaconBlockHeader attested_header = 1;
+    SyncCommittee next_sync_committee = 2;
+    repeated bytes next_sync_committee_branch = 3 [(ethereum.eth.ext.ssz_size) = "next_sync_committee_branch.depth,32"];
+    v1.BeaconBlockHeader finalized_header = 4;
+    repeated bytes finality_branch = 5 [(ethereum.eth.ext.ssz_size) = "finality_branch.depth,32"];
+    v1.SyncAggregate sync_aggregate = 6;
+    uint64 signature_slot = 7 [(ethereum.eth.ext.cast_type) = "github.com/prysmaticlabs/prysm/v4/consensus-types/primitives.Slot"];
+}
+
+message LightClientFinalityUpdateResponse {
```
Comment (rkapka) [inline proto/eth/v2/beacon_lightclient.proto:49]: This type can be defined under `rpc/eth/lightclient` because it doesn't need ssz

File: proto/eth/v2/beacon_lightclient.proto [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+message LightClientFinalityUpdate {
+    v1.BeaconBlockHeader attested_header = 1;
+    v1.BeaconBlockHeader finalized_header = 2;
+    repeated bytes finality_branch = 3 [(ethereum.eth.ext.ssz_size) = "finality_branch.depth,32"];
+    v1.SyncAggregate sync_aggregate = 4;
+    uint64 signature_slot = 5 [(ethereum.eth.ext.cast_type) = "github.com/prysmaticlabs/prysm/v4/consensus-types/primitives.Slot"];
+}
+
+message LightClientOptimisticUpdateResponse {
```
Comment (rkapka) [inline proto/eth/v2/beacon_lightclient.proto:62]: This type can be defined under `rpc/eth/lightclient` because it doesn't need ssz

File: beacon-chain/rpc/eth/helpers/error_handling.go (hunk 3 above)
Comment (nicopernas) [reply beacon-chain/rpc/eth/helpers/error_handling.go:85]: Yes! that function was not there when I wrote the initial refactor. Will use that one instead.

File: beacon-chain/rpc/eth/helpers/lightclient/lightclient.go (hunk 4 above)
Comment (nicopernas) [reply beacon-chain/rpc/eth/helpers/lightclient/lightclient.go:29]: I did remove the grpc status and codes from the handlers but these ones I missed. Will replace them.

File: proto/eth/v2/custom.go (hunk 5 above)
Comment (nicopernas) [reply proto/eth/v2/custom.go:10]: They are used by the light client store later in the series.

File: proto/eth/v2/custom.go (hunk 6 above)
Comment (nicopernas) [reply proto/eth/v2/custom.go:25]: Also used by the light client store.

File: proto/eth/v2/beacon_lightclient.proto (hunk 7 above)
Comment (nicopernas) [reply proto/eth/v2/beacon_lightclient.proto:49]: This type is used in the http streaming api (see [here](https://github.com/polymerdao/prysm/blob/polymer/lightclient-5/beacon-chain/rpc/eth/events/events.go#L217))  and `streamData()` expects a `proto.Message`.

**Comment by notbdu** [comment]
> Hi @nicopernas . We are currently in the process of switching our APIs from gRPC to HTTP. The first step is this tracking issue: #12635. We want to ultimately remove protobufs and use "pure" Go structs in our whole codebase, along with HTTP handlers serving API requests. That being said, we currently still use protobufs in our core code and for ssz generation.
> 
> I see that none of your types are added to the `ssz_gen_marshal` bazel rule, which makes me think that it is not necessary to define them as protos. On the other hand, all light client endpoints in the Beacon API specs allow an `Accept` header, meaning they can be returned in SSZ format, so it seems that protos are the best choice.
> 
> I think it would be most future-proof to split all types into 3 groups:
> 
> 1. API-only structs defined in `/beacon-chain/rpc/beacon/light_client_structs.go`
> 2. Structs that will be used in core light client code and don't need SSZ serialization, defined under `/consensus-types/light-client` (there might not be any)
> 3. Structs that will be used in core light client code and need SSZ serialization, defined in `proto/eth/v2/beacon_lightclient.proto`

What's the reason for the migration from gRPC to HTTP? Is it an issue w/ most downstream clients using HTTP and the perf overhead of the grpc gateway or something like that?

File: beacon-chain/rpc/eth/helpers/lightclient/lightclient.go (hunk 4 above)
Comment (nicopernas) [reply beacon-chain/rpc/eth/helpers/lightclient/lightclient.go:29]: After replacing the grpc types with `core.RpcError` I got myself into a circular dependency:
- the LC helper depends on rpc.core
- rpc.core depends on blockchain
- blockchain depends on LC helper (this is added later in the series)

Easiest thing I can think of is moving the rpc/core/errors.go to a "submodule" like `rpc.core.errors`. That will only affect my lc helper and `beacon-chain/rpc/core/validator.go`. 

What do you think?

File: proto/eth/v2/beacon_lightclient.proto (hunk 7 above)
Comment (rkapka) [reply proto/eth/v2/beacon_lightclient.proto:49]: I think we should differentiate between proto messages used in events and HTTP response structs that don't need to be proto messages. After all they are used in two different contexts. I would remove the `Response` suffix from protos and create the response structs next to HTTP handlers.

File: beacon-chain/rpc/eth/helpers/lightclient/lightclient.go (hunk 4 above)
Comment (rkapka) [reply beacon-chain/rpc/eth/helpers/lightclient/lightclient.go:29]: RPC is the outer layer of the system. It should not depend on the blockchain package. If your design requires this, then these helper functions should also be under blockchain. I think a new `blockchain/lightclient` package would make most sense. I haven't reviewed the logic of these functions yet, but they don't seem to be closely connected with any rpc functionality. Such a change would also mean that regular go errors will be returned from these functions.

**Comment by rkapka** [comment]
> > Hi @nicopernas . We are currently in the process of switching our APIs from gRPC to HTTP. The first step is this tracking issue: #12635. We want to ultimately remove protobufs and use "pure" Go structs in our whole codebase, along with HTTP handlers serving API requests. That being said, we currently still use protobufs in our core code and for ssz generation.
> > I see that none of your types are added to the `ssz_gen_marshal` bazel rule, which makes me think that it is not necessary to define them as protos. On the other hand, all light client endpoints in the Beacon API specs allow an `Accept` header, meaning they can be returned in SSZ format, so it seems that protos are the best choice.
> > I think it would be most future-proof to split all types into 3 groups:
> > 
> > 1. API-only structs defined in `/beacon-chain/rpc/beacon/light_client_structs.go`
> > 2. Structs that will be used in core light client code and don't need SSZ serialization, defined under `/consensus-types/light-client` (there might not be any)
> > 3. Structs that will be used in core light client code and need SSZ serialization, defined in `proto/eth/v2/beacon_lightclient.proto`
> 
> What's the reason for the migration from gRPC to HTTP? Is it an issue w/ most downstream clients using HTTP and the perf overhead of the grpc gateway or something like that?

Hey! Yes, you are basically correct. Because of how Beacon APIs are specified, we require additional middleware to translate between HTTP and gRPC, which adds maintenance cost and significant overhead.

File: beacon-chain/rpc/eth/helpers/lightclient/lightclient.go (hunk 4 above)
Comment (nicopernas) [reply beacon-chain/rpc/eth/helpers/lightclient/lightclient.go:29]: The LC helpers introduced in this PR do not depend on the blockchain package. The blockchain package will depend on the LC helpers in a future PR. 

However, `rpc.core` does depend on the `blockchain` package and that's where the circular dependency comes from. 

Not adding a dependency from the LC helpers on `rpc.core` will solve the issue. I think using regular errors will do the trick regardless of where the LC helpers live.

File: proto/eth/v2/beacon_lightclient.proto (hunk 7 above)
Comment (nicopernas) [reply proto/eth/v2/beacon_lightclient.proto:49]: I can remove the `Response` suffix, no prob. 

These types are only used in the events and not by the HTTP handlers. The objects that populate the HTTP handlers' responses are already defined in go structs next to the handlers as you initially suggested. See https://github.com/polymerdao/prysm/blob/polymer/lightclient-5/beacon-chain/rpc/eth/light-client/structs.go

File: beacon-chain/rpc/eth/helpers/lightclient/lightclient.go (hunk 4 above)
Comment (rkapka) [reply beacon-chain/rpc/eth/helpers/lightclient/lightclient.go:29]: I want to avoid the blockchain package depending on anything in the rpc package, if possible.

File: proto/eth/v2/beacon_lightclient.proto (hunk 7 above)
Comment (nicopernas) [reply proto/eth/v2/beacon_lightclient.proto]: I ended up replacing the `Response` suffix with `WithVersion`. The idea is that we wanted to follow the spec and have these methods: [`create_light_client_finality_update`](https://github.com/ethereum/consensus-specs/blob/3d235740e5f1e641d3b160c8688f26e7dc5a1894/specs/altair/light-client/full-node.md#create_light_client_finality_update) and [`create_light_client_optimistic_update`](https://github.com/ethereum/consensus-specs/blob/3d235740e5f1e641d3b160c8688f26e7dc5a1894/specs/altair/light-client/full-node.md#create_light_client_optimistic_update). The result is having the structs we have: `LightClientOptimisticUpdate` and `LightClientFinalityUpdate`. Adding a version to these "goes against" the spec. 

The other approach that we considered was deviating a bit from the spec and have more "efficient" types. For example, we could `LightClientUpdate` to represent both final and optimistic updates. Note that as of now, `LightClientUpdate` is a super set of `LightClientOptimisticUpdate` and `LightClientFinalityUpdate`. The single type could be returned from both finality and optimistic endpoints nullifying the correspondent fields in each case. While we are at it, we could throw the version in there too. 

Again, since we wanted to stay closer to the spec we went for the approach proposed in this PR. However, there's an argument to be made in favour of the second approach.

File: beacon-chain/blockchain/lightclient.go [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+	}
+
+	// assert sum(block.message.body.sync_aggregate.sync_committee_bits) >= MIN_SYNC_COMMITTEE_PARTICIPANTS
+	syncAggregate, err := block.Block().Body().SyncAggregate()
+	if err != nil {
+		return nil, fmt.Errorf("could not get sync aggregate: %v", err)
+	}
+
+	if syncAggregate.SyncCommitteeBits.Count() < config.MinSyncCommitteeParticipants {
+		return nil, fmt.Errorf("invalid sync committee bits count: %d", syncAggregate.SyncCommitteeBits.Count())
```
Comment (rkapka) [inline beacon-chain/blockchain/lightclient.go:37]:
```suggestion
		return nil, fmt.Errorf("invalid sync committee bits count %d", syncAggregate.SyncCommitteeBits.Count())
```

File: beacon-chain/blockchain/lightclient.go [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+		return nil, fmt.Errorf("could not get header root: %v", err)
+	}
+
+	blockRoot, err := block.Block().HashTreeRoot()
+	if err != nil {
+		return nil, fmt.Errorf("could not get block root: %v", err)
+	}
+
+	if headerRoot != blockRoot {
+		return nil, fmt.Errorf("invalid header root: %v", headerRoot)
```
Comment (rkapka) [inline beacon-chain/blockchain/lightclient.go:64]:
```suggestion
		return nil, fmt.Errorf("header root %#x not equal to block root %#x", headerRoot, blockRoot)
```

File: beacon-chain/blockchain/lightclient.go [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("could not get attested state root: %v", err)
+	}
+	attestedHeader.StateRoot = attestedStateRoot[:]
+
+	// assert hash_tree_root(attested_header) == block.message.parent_root
+	attestedHeaderRoot, err := attestedHeader.HashTreeRoot()
+	if err != nil || attestedHeaderRoot != block.Block().ParentRoot() {
+		return nil, fmt.Errorf("invalid attested header root: %v", attestedHeaderRoot)
+	}
```
Comment (rkapka) [inline beacon-chain/blockchain/lightclient.go:86]: let's split this into two error messages - one for `err != nil` and one for non-equal roots

File: beacon-chain/blockchain/lightclient.go [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+
+func NewLightClientOptimisticUpdateFromBeaconState(
+	ctx context.Context,
+	config *params.BeaconChainConfig,
+	state state.BeaconState,
+	block interfaces.ReadOnlySignedBeaconBlock,
+	attestedState state.BeaconState) (*ethpbv2.LightClientUpdate, error) {
+	// assert compute_epoch_at_slot(attested_state.slot) >= ALTAIR_FORK_EPOCH
+	attestedEpoch := types.Epoch(uint64(attestedState.Slot()) / uint64(config.SlotsPerEpoch))
+	if attestedEpoch < types.Epoch(config.AltairForkEpoch) {
```
Comment (rkapka) [inline beacon-chain/blockchain/lightclient.go:26]: You don't have to pass the config as a parameter. Use `params.BeaconConfig()` instead

File: beacon-chain/blockchain/lightclient.go [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+	} else {
+		finalizedHeader = &ethpbv1.BeaconBlockHeader{
+			Slot:          0,
+			ProposerIndex: 0,
+			ParentRoot:    make([]byte, 32),
+			StateRoot:     make([]byte, 32),
+			BodyRoot:      make([]byte, 32),
+		}
+
+		finalityBranch = make([][]byte, 6)
```
Comment (rkapka) [inline beacon-chain/blockchain/lightclient.go:177]: Please extract `6` to a constant with a meaningful name

File: beacon-chain/blockchain/lightclient.go [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+			}
+			finalizedHeader = migration.V1Alpha1SignedHeaderToV1(tempFinalizedHeader).GetMessage()
+
+			finalizedHeaderRoot, err := finalizedHeader.HashTreeRoot()
+			if err != nil {
+				return nil, fmt.Errorf("could not get finalized header root: %v", err)
+			}
+
+			if finalizedHeaderRoot != bytesutil.ToBytes32(attestedState.FinalizedCheckpoint().Root) {
+				return nil, fmt.Errorf("invalid finalized header root: %v", finalizedHeaderRoot)
```
Comment (rkapka) [inline beacon-chain/blockchain/lightclient.go:147]:
```suggestion
				return nil, fmt.Errorf("finalized header root %#x not equal to attested finalized checkpoint root %#x", finalizedHeaderRoot, bytesutil.ToBytes32(attestedState.FinalizedCheckpoint().Root)
```

**Review by rkapka — CHANGES_REQUESTED** [review]
Both functions in `beacon-chain/blockchain/lightclient.go` don't have any tests. Please add a test for the happy case as a minimum.

**Comment by rkapka** [comment]
One thing I don't see in the PR is the following part of the `create_light_client_update` function from the spec:
```
# `next_sync_committee` is only useful if the message is signed by the current sync committee
    if update_attested_period == update_signature_period:
        update.next_sync_committee = attested_state.next_sync_committee
        update.next_sync_committee_branch = compute_merkle_proof_for_state(
            attested_state, NEXT_SYNC_COMMITTEE_INDEX)
```
Is this on purpose?

**Comment by nicopernas** [comment]
> One thing I don't see in the PR is the following part of the `create_light_client_update` function from the spec:
> 
> ```
> # `next_sync_committee` is only useful if the message is signed by the current sync committee
>     if update_attested_period == update_signature_period:
>         update.next_sync_committee = attested_state.next_sync_committee
>         update.next_sync_committee_branch = compute_merkle_proof_for_state(
>             attested_state, NEXT_SYNC_COMMITTEE_INDEX)
> ```
> 
> Is this on purpose?

The `create_light_client_update` implementation will be added by the next PR in the series. See https://github.com/polymerdao/prysm/blob/polymer/lightclient-5/beacon-chain/rpc/eth/light-client/helpers.go#L95

I have resolved your latest comments just so I could track the ones I had left to fix. Will be working on the happy path testes now and send a new update.

**Comment by rkapka** [comment]
Ah, ok. So these two functions are not implementations of any particular spec function?

**Comment by nicopernas** [comment]
They are actually, it's just that we organised the code in a funny way. Not sure why though...

* `NewLightClientFinalityUpdateFromBeaconState` implements [create_light_client_finality_update](https://github.com/ethereum/consensus-specs/blob/3d235740e5f1e641d3b160c8688f26e7dc5a1894/specs/altair/light-client/full-node.md#create_light_client_finality_update)
* `NewLightClientOptimisticUpdateFromBeaconState` implements [create_light_client_optimistic_update](https://github.com/ethereum/consensus-specs/blob/3d235740e5f1e641d3b160c8688f26e7dc5a1894/specs/altair/light-client/full-node.md#create_light_client_optimistic_update)

See how `sendLightClientFinalityUpdate()` and `sendLightClientOptimisticUpdate()` use them in [process_block.go](https://github.com/polymerdao/prysm/blob/polymer/lightclient-5/beacon-chain/blockchain/process_block.go#L95) and you'll see what I mean. 

I will refactor and merge these to avoid this type of confusions.

**Comment by rkapka** [comment]
I was actually asking about the `XXXFromBeaconState` functions. They look very similar to `create_light_client_update`, which confused me.

**Comment by rkapka** [comment]
You will have to run `bazel run //:gazelle -- fix` to satisfy the linter

**Comment by rkapka** [comment]
Looks like the branch has some conflicts. Happy to merge this in once it's resolved

**Comment by nicopernas** [comment]
Thanks! I just rebased from latest `develop` and fixed the conflicts.

File: beacon-chain/blockchain/lightclient_test.go [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+	l.checkAttestedHeader(update)
+
+	zeroHash := params.BeaconConfig().ZeroHash
+	require.NotNil(t, update.FinalizedHeader, "Finalized header is nil")
+	require.Equal(t, primitives.Slot(0), update.FinalizedHeader.Slot, "Finalized header slot is not zero")
+	require.Equal(t, primitives.ValidatorIndex(0), update.FinalizedHeader.ProposerIndex, "Finalized header proposer index is not zero")
+	require.DeepSSZEqual(t, zeroHash, update.FinalizedHeader.ParentRoot, "Finalized header parent root is not zero")
+	require.DeepSSZEqual(t, zeroHash, update.FinalizedHeader.StateRoot, "Finalized header state root is not zero")
+	require.DeepSSZEqual(t, zeroHash, update.FinalizedHeader.BodyRoot, "Finalized header body root is not zero")
+	require.Equal(t, 6, len(update.FinalityBranch), "Invalid finality branch leaves")
```
Comment (rkapka) [inline beacon-chain/blockchain/lightclient_test.go:156]:
```suggestion
	require.Equal(t, finalityBranchNumOfLeaves, len(update.FinalityBranch), "Invalid finality branch leaves")
```

**Review by rkapka — COMMENTED** [review]
1 nitpick

**Comment by nicopernas** [comment]
> 1 nitpick

Fixed and rebased from latest `develop`

**Review by rkapka — APPROVED** [review]

**Comment by rkapka** [comment]
> > 1 nitpick
> 
> Fixed and rebased from latest `develop`

Thanks a lot for this great work, I approved the PR. Looks like something else got merged in the meantime and the branch is out of date again :(

**Comment by nicopernas** [comment]
> > > 1 nitpick
> > 
> > 
> > Fixed and rebased from latest `develop`
> 
> Thanks a lot for this great work, I approved the PR. Looks like something else got merged in the meantime and the branch is out of date again :(

How about now?
